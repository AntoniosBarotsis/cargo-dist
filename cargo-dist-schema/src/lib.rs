use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
// FIXME: ideally these would be UTF8PathBufs but JsonSchema doesn't support (yet?)
use std::path::PathBuf;

/// The final report of cargo-dist
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct DistManifest {
    /// App releases we're distributing
    pub releases: Vec<Release>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct Release {
    /// The name of the app
    pub app_name: String,
    /// The version of the app
    // FIXME: should be a Version but JsonSchema doesn't support (yet?)
    pub app_version: String,
    /// The artifacts for this release (zips, debuginfo, metadata...)
    pub artifacts: Vec<Artifact>,
}

/// A distributable bundle that's part of a Release
///
/// i.e. a zip or installer
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct Artifact {
    /// The unique name of the artifact (e.g. `myapp-v1.0.0-x86_64-pc-windows-msvc.zip`)
    pub name: String,
    /// The kind of artifact this is (e.g. "exectuable-zip")
    #[serde(flatten)]
    pub kind: ArtifactKind,
    /// The target triple of the bundle
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_triple: Option<String>,
    /// The location of the artifact on the local system
    #[serde(skip_serializing_if = "Option::is_none")]
    pub path: Option<PathBuf>,
    /// Assets included in the bundle (like executables)
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub assets: Vec<Asset>,
}

/// An asset contained in an artifact (executable, license, etc.)
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct Asset {
    /// The high-level name of the asset
    pub name: String,
    /// The path of the asset relative to the root of the artifact
    pub path: PathBuf,
    /// The kind of asset this is
    #[serde(flatten)]
    pub kind: AssetKind,
}

/// An artifact included in a Distributable
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(tag = "kind")]
#[non_exhaustive]
pub enum AssetKind {
    /// An executable artifact
    #[serde(rename = "executable")]
    Executable(ExecutableAsset),
    /// A README file
    #[serde(rename = "readme")]
    Readme,
    /// A LICENSE file
    #[serde(rename = "license")]
    License,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(tag = "kind")]
pub enum ArtifactKind {
    /// A zip or a tarball
    #[serde(rename = "executable-zip")]
    ExecutableZip,
    /// Standalone Symbols/Debuginfo for a build
    #[serde(rename = "symbols")]
    Symbols,
    /// Machine-readable metadata
    #[serde(rename = "dist-metadata")]
    DistMetadata,
    /// Installer
    #[serde(rename = "installer")]
    Installer,
}

/// An executable artifact (exe/binary)
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct ExecutableAsset {
    /// The name of the Artifact containing symbols for this executable
    #[serde(skip_serializing_if = "Option::is_none")]
    pub symbols_artifact: Option<String>,
}

impl DistManifest {
    pub fn new(releases: Vec<Release>) -> Self {
        Self { releases }
    }
}

#[test]
fn emit() {
    let schema = schemars::schema_for!(DistManifest);
    insta::assert_snapshot!(serde_json::to_string_pretty(&schema).unwrap());
}
