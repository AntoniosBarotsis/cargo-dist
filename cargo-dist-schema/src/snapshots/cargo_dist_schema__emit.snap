---
source: cargo-dist-schema/src/lib.rs
expression: json_schema
---
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "DistManifest",
  "description": "A report of the releases and artifacts that cargo-dist generated",
  "type": "object",
  "properties": {
    "dist_version": {
      "description": "The version of cargo-dist that generated this",
      "type": [
        "string",
        "null"
      ]
    },
    "releases": {
      "description": "App releases we're distributing",
      "type": "array",
      "items": {
        "$ref": "#/definitions/Release"
      }
    }
  },
  "definitions": {
    "Artifact": {
      "description": "A distributable bundle that's part of a Release\n\ni.e. a zip or installer",
      "type": "object",
      "oneOf": [
        {
          "description": "A zip or a tarball",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "executable-zip"
              ]
            }
          }
        },
        {
          "description": "Standalone Symbols/Debuginfo for a build",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "symbols"
              ]
            }
          }
        },
        {
          "description": "Machine-readable metadata",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "dist-metadata"
              ]
            }
          }
        },
        {
          "description": "Installer",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "installer"
              ]
            }
          }
        },
        {
          "description": "Unknown to this version of cargo-dist-schema\n\nThis is a fallback for forward/backward-compat",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "unknown"
              ]
            }
          }
        }
      ],
      "required": [
        "name"
      ],
      "properties": {
        "assets": {
          "description": "Assets included in the bundle (like executables and READMEs)",
          "type": "array",
          "items": {
            "$ref": "#/definitions/Asset"
          }
        },
        "install_hint": {
          "description": "A string describing how to install this",
          "type": [
            "string",
            "null"
          ]
        },
        "name": {
          "description": "The unique name of the artifact (e.g. `myapp-v1.0.0-x86_64-pc-windows-msvc.zip`)",
          "type": "string"
        },
        "path": {
          "description": "The location of the artifact on the local system",
          "type": [
            "string",
            "null"
          ]
        },
        "target_triples": {
          "description": "The target triple of the bundle",
          "type": "array",
          "items": {
            "type": "string"
          }
        }
      }
    },
    "Asset": {
      "description": "An asset contained in an artifact (executable, license, etc.)",
      "type": "object",
      "oneOf": [
        {
          "description": "An executable artifact",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "executable"
              ]
            },
            "symbols_artifact": {
              "description": "The name of the Artifact containing symbols for this executable",
              "type": [
                "string",
                "null"
              ]
            }
          }
        },
        {
          "description": "A README file",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "readme"
              ]
            }
          }
        },
        {
          "description": "A LICENSE file",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "license"
              ]
            }
          }
        },
        {
          "description": "A CHANGELOG or RELEASES file",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "changelog"
              ]
            }
          }
        },
        {
          "description": "Unknown to this version of cargo-dist-schema\n\nThis is a fallback for forward/backward-compat",
          "type": "object",
          "required": [
            "kind"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "unknown"
              ]
            }
          }
        }
      ],
      "properties": {
        "name": {
          "description": "The high-level name of the asset",
          "type": [
            "string",
            "null"
          ]
        },
        "path": {
          "description": "The path of the asset relative to the root of the artifact",
          "type": [
            "string",
            "null"
          ]
        }
      }
    },
    "Release": {
      "description": "A Release of an Application",
      "type": "object",
      "required": [
        "app_name",
        "app_version"
      ],
      "properties": {
        "app_name": {
          "description": "The name of the app",
          "type": "string"
        },
        "app_version": {
          "description": "The version of the app",
          "type": "string"
        },
        "artifacts": {
          "description": "The artifacts for this release (zips, debuginfo, metadata...)",
          "type": "array",
          "items": {
            "$ref": "#/definitions/Artifact"
          }
        }
      }
    }
  }
}
